addons/pid/fnc_update.sqf

```diff
@@ -0,0 +1,103 @@
+#include "script_component.hpp"
+/* ----------------------------------------------------------------------------
+Function: CBA_pid_fnc_update
+
+Description:
+    Updates a PID controller and returns the controller output.
+
+Parameters:
+    _pid    - a pid controller <LOCATION>
+    _value  - the measured value for this update <NUMBER>
+    _time   - the time for which the value was measured <NUMBER>
+              (Default: CBA_missionTime)
+
+Returns:
+    The output of the controller <NUMBER>
+
+Examples:
+    (begin example)
+        private _delta = [_pid, 6, CBA_missionTime] call CBA_pid_fnc_update;
+    (end)
+
+Author:
+    tcvm
+---------------------------------------------------------------------------- */
+SCRIPT(update);
+params [
+    ["_pid", locationNull, [locationNull]],
+    ["_value", 0, [0]],
+    ["_time", CBA_missionTime, [0]]
+];
+
+if (isNull _pid) exitWith { 0 };
+private _error = [_value, _pid getVariable [QGVAR(setpoint), 0]] call (_pid getVariable [QGVAR(errorFunction), FUNC(error_linear)]);
+private _history = _pid getVariable [QGVAR(history), []];
+private _maxHistoryLength = _pid getVariable [QGVAR(historyLength), DEFAULT_HISTORY_LENGTH];
+
+if (_history isNotEqualTo []) then {
+    // We need a continuous function, so if two measurements are for the same measured time we ignore the new measurements
+    // This also gets rid of any potential divide-by-zero errors due to the stride being 0
+    // The inverse must also be monotonic, so we assume time is always increasing
+    (_history select -1) params ["_x"];
+    if (_time - _x > 0) then {
+        _history pushBack [_time, _error];
+    };
+} else {
+    _history pushBack [_time, _error];
+};
+
+if (count _history > _maxHistoryLength) then {
+    _history deleteAt 0;
+};
+_pid setVariable [QGVAR(history), _history];
+
+private _derivative = 0;
+switch (true) do {
+    case (count _history == 2): {
+        private _x0 = _history select 0;
+        private _x1 = _history select 1;
+
+        private _stride = (_x1 select 0) - (_x0 select 0);
+        if (_stride == 0) then { break };
```

Comment:
```suggestion
        if (_stride == 0) exitWith {};
```
not in a loop so the `break` may hit something unexpected
example
```
[{
    systemChat "A"; if (true) then { break }; systemChat "unreachable";
}] call CBA_fnc_addPerFrameHandler;
[{
    systemChat "B"; // also never reached because `break` hit some upper pfeh-internal loop
}] call CBA_fnc_addPerFrameHandler;
```